File: app/src/main/java/no/nav/tilbakemeldingsmottak/bigquery/serviceklager/ServiceklageSlettScheduled.java
package no.nav.tilbakemeldingsmottak.bigquery.serviceklager;

import com.google.cloud.bigquery.BigQuery;
import com.google.cloud.bigquery.QueryJobConfiguration;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.tilbakemeldingsmottak.repository.ServiceklageRepository;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Profile;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
@Slf4j
@Profile("nais | itest")
public class ServiceklageSlettScheduled {
    private final BigQuery bigQueryClient;
    private final ServiceklagerBigQuery serviceklagerBigQuery;
    private final ServiceklageRepository serviceklageRepository;


    @Value("${big_query_dataset}")
    private String dataset;

    @Value("${gcp_team_project_id}")
    private String projectId;

    @Value("${cron.slettBigQueryServiceKlagerEldreEnn}")
    private String slettBigQueryServiceKlagerEldreEnn;


    private String slettEldreEnnQuery(String datoFelt, String eventType) {
        return String.format(
                "DELETE FROM `%s.%s.%s` " +
                        "WHERE TIMESTAMP_DIFF(CURRENT_TIMESTAMP(), %s, DAY) >= %s " +
                        "AND event_type='%s'",
                projectId, dataset, ServiceklagerBigQuery.TABLE_NAME,
                datoFelt, slettBigQueryServiceKlagerEldreEnn,
                eventType
        );
    }

    @Scheduled(cron = "${cron.slettBigQueryServiceKlagerScheduled}")
    public void slettServiceKlager() {
        try {
            log.info("Sletter serviceklager eldre enn {} dager", slettBigQueryServiceKlagerEldreEnn);

            // Slett alle opprettede serviceklager som er eldre enn 1 uke
            var slettOpprettedeServiceklager = slettEldreEnnQuery("opprettet_dato", ServiceklageEventTypeEnum.OPPRETT_SERVICEKLAGE.value);
            log.info("Query for sletting av opprettede serviceklager: {}", slettOpprettedeServiceklager);

            var slettOpprettedeServiceklagerQueryConfig = QueryJobConfiguration.newBuilder(slettOpprettedeServiceklager).build();
            bigQueryClient.query(slettOpprettedeServiceklagerQueryConfig);

            // Slett alle klassifiserte serviceklager som er eldre enn 1 uke
            var slettKlassifiserteServiceklager = slettEldreEnnQuery("avsluttet_dato", ServiceklageEventTypeEnum.KLASSIFISER_SERVICEKLAGE.value);
            log.info("Query for sletting av klassifiserte serviceklager: {}", slettOpprettedeServiceklager);

            var slettKlassifiserteServiceklagerQueryConfig = QueryJobConfiguration.newBuilder(slettKlassifiserteServiceklager).build();
            bigQueryClient.query(slettKlassifiserteServiceklagerQueryConfig);

        } catch (Exception e) {
            log.error("Kunne ikke slette Big Query serviceklager eldre enn {} dager", slettBigQueryServiceKlagerEldreEnn, e);
        }
    }

    @Scheduled(cron = "0 0 14 9 AUG ?")
    public void oppdaterServiceklager() {
        // Skal ikke kjøres årlig. Har blitt brukt til "manuelle" oppdateringer
        if (LocalDateTime.now().getYear() == 2023) {
            var datoFra = LocalDateTime.now().minusMonths(4);

            try {
                log.info("Oppdaterer serviceklager fra {}", datoFra);

                var serviceklager = serviceklageRepository.findAllByOpprettetDatoAfterOrAvsluttetDatoAfter(datoFra, datoFra);

                for (var serviceklage : serviceklager) {
                    serviceklagerBigQuery.insertServiceklage(serviceklage, ServiceklageEventTypeEnum.OPPDATER_SERVICEKLAGE);
                }

                log.info("Oppdatert serviceklager fra {}", datoFra);

            } catch (Exception ex) {
                log.error("Kunne ikke oppdatere Big Query serviceklager fra {}", datoFra, ex);
            }
        }
    }


}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/bigquery/serviceklager/ServiceklageSlettScheduled.kt
package no.nav.tilbakemeldingsmottak.bigquery.serviceklager

import com.google.cloud.bigquery.BigQuery
import com.google.cloud.bigquery.QueryJobConfiguration
import no.nav.tilbakemeldingsmottak.repository.ServiceklageRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Profile
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime

@Service
@Profile("nais | itest")
class ServiceklageSlettScheduled(
    private val bigQueryClient: BigQuery,
    private val serviceklagerBigQuery: ServiceklagerBigQuery,
    private val serviceklageRepository: ServiceklageRepository
) {

    @Value("\${big_query_dataset}")
    private val dataset: String? = null

    @Value("\${gcp_team_project_id}")
    private val projectId: String? = null

    @Value("\${cron.deleteBigQueryServiceKlagerOlderThan}")
    private val deleteBigQueryServiceKlagerOlderThan: String? = null

    @Value("\${cron.deleteServiceKlagerOlderThan}")
    private val deleteServiceKlagerOlderThan: String? = null

    private val logger = LoggerFactory.getLogger(javaClass)
    private fun slettEldreEnnQuery(datoFelt: String, eventType: String): String {
        return String.format(
            "DELETE FROM `%s.%s.%s` " +
                    "WHERE TIMESTAMP_DIFF(CURRENT_TIMESTAMP(), %s, DAY) >= %s " +
                    "AND event_type='%s'",
            projectId, dataset, ServiceklagerBigQuery.TABLE_NAME,
            datoFelt, deleteBigQueryServiceKlagerOlderThan,
            eventType
        )
    }

    @Scheduled(cron = "\${cron.deleteServiceKlagerScheduled}")
    @Transactional
    fun deleteServiceKlager() {
        try {
            logger.info("Skal slette serviceklager fra databasen eldre enn {} dager", deleteServiceKlagerOlderThan)
            val cutoffDate = LocalDateTime.now().minusDays(deleteServiceKlagerOlderThan?.toLong() ?: 90)
            serviceklageRepository.deleteServiceklageOlderThan(cutoffDate)
            logger.info("Slettet serviceklager fra databasen eldre enn {} dager", deleteServiceKlagerOlderThan)
        } catch (e: Exception) {
            logger.error("Kunne ikke slette serviceklager eldre enn {} dager", deleteServiceKlagerOlderThan, e)
        }
    }

    @Scheduled(cron = "\${cron.deleteBigQueryServiceKlagerScheduled}")
    fun deleteBigQueryServiceKlager() {
        try {
            logger.info(
                "Skal slette serviceklager fra big query eldre enn {} dager",
                deleteBigQueryServiceKlagerOlderThan
            )

            // Slett alle opprettede serviceklager som er eldre enn 1 uke
            val slettOpprettedeServiceklager =
                slettEldreEnnQuery("opprettet_dato", ServiceklageEventType.OPPRETT_SERVICEKLAGE.name)

            logger.info("Query for sletting av opprettede serviceklager: {}", slettOpprettedeServiceklager)

            val slettOpprettedeServiceklagerQueryConfig =
                QueryJobConfiguration.newBuilder(slettOpprettedeServiceklager).build()
            bigQueryClient.query(slettOpprettedeServiceklagerQueryConfig)

            // Slett alle klassifiserte serviceklager som er eldre enn 1 uke
            val slettKlassifiserteServiceklager =
                slettEldreEnnQuery("avsluttet_dato", ServiceklageEventType.KLASSIFISER_SERVICEKLAGE.name)

            logger.info("Query for sletting av klassifiserte serviceklager: {}", slettOpprettedeServiceklager)

            val slettKlassifiserteServiceklagerQueryConfig =
                QueryJobConfiguration.newBuilder(slettKlassifiserteServiceklager).build()
            bigQueryClient.query(slettKlassifiserteServiceklagerQueryConfig)

            logger.info("Slettet serviceklager fra big query eldre enn {} dager", deleteBigQueryServiceKlagerOlderThan)

        } catch (e: Exception) {
            logger.error(
                "Kunne ikke slette Big Query serviceklager eldre enn {} dager",
                deleteBigQueryServiceKlagerOlderThan,
                e
            )
        }
    }

    @Scheduled(cron = "0 0 14 9 AUG ?")
    fun oppdaterServiceklager() {
        // Skal ikke kjøres årlig. Har blitt brukt til "manuelle" oppdateringer
        if (LocalDateTime.now().year == 2023) {
            val datoFra = LocalDateTime.now().minusMonths(4)
            try {
                logger.info("Oppdaterer serviceklager fra {}", datoFra)
                val serviceklager =
                    serviceklageRepository.findAllByOpprettetDatoAfterOrAvsluttetDatoAfter(datoFra, datoFra)
                for (serviceklage in serviceklager) {
                    serviceklagerBigQuery.insertServiceklage(
                        serviceklage,
                        ServiceklageEventType.OPPDATER_SERVICEKLAGE
                    )
                }
                logger.info("Oppdatert serviceklager fra {}", datoFra)
            } catch (ex: Exception) {
                logger.error("Kunne ikke oppdatere Big Query serviceklager fra {}", datoFra, ex)
            }
        }
    }
}
